### PR TITLE
KT-33783 KDoc: Detail when a range is empty

### PR DESCRIPTION
**Pull request by shadrina**
[KT-33783](https://youtrack.jetbrains.com/issue/KT-33783)

@p7nov, please take a look.

### CODE REVIEW

File: core/builtins/src/kotlin/Progressions.kt [hunk 1]
```diff
@@ -98,7 +103,12 @@ public open class IntProgression
 
     override fun iterator(): IntIterator = IntProgressionIterator(first, last, step)
 
-    /** Checks if the progression is empty. */
+    /**
+     * Checks if the progression is empty.
+     *
+     * Increasing progression is empty if its start value is greater than the end value.
+     * Decreasing progression is empty if its start value is less than the end value.
```
Comment (ilya-g) [inline core/builtins/src/kotlin/Progressions.kt:110]: A progression doesn't have start and end values. It has the properties returning `first` and `last` elements. The difference is explained here: https://kotlinlang.org/docs/ranges.html#progression

**Review by ilya-g — CHANGES_REQUESTED** [review]

File: core/builtins/src/kotlin/Progressions.kt (hunk 1 above)
Comment (shadrina) [reply core/builtins/src/kotlin/Progressions.kt]: Changed the wording to `start`/`end` elements

File: core/builtins/src/kotlin/Progressions.kt [hunk 2]
```diff
@@ -41,7 +41,12 @@ public open class CharProgression
 
     override fun iterator(): CharIterator = CharProgressionIterator(first, last, step)
 
-    /** Checks if the progression is empty. */
+    /**
+     * Checks if the progression is empty.
+     *
+     * Increasing progression is empty if its start value is greater than the end value.
```
[the remark this answers is not in the document]
Comment (shadrina) [reply core/builtins/src/kotlin/Progressions.kt]: Clarified `isEmpty` documentation for ranges, kept the wording (start/end values)

[the remark this answers is not in the document]
Comment (ilya-g) [reply core/builtins/src/kotlin/Progressions.kt]: Also please in `core/builtins/src/kotlin/Range.kt`

**Review by ilya-g — COMMENTED** [review]
Now the wording is fine for me, please check the grammar with @p7nov

**Review by p7nov — APPROVED** [review]
👍

File: generators/builtins/unsignedTypes.kt [hunk 3]
```diff
@@ -579,7 +584,12 @@ internal constructor(
 
     override fun iterator(): ${elementType}Iterator = ${elementType}ProgressionIterator(first, last, step)
 
-    /** Checks if the progression is empty. */
+    /** 
+     * Checks if the progression is empty.
+     
+     * Progression with positive step is empty if its first element is greater than the last element.
```
Comment (ilya-g) [inline generators/builtins/unsignedTypes.kt:590]: Either here should be an article as in Progression docs, "... with _a_ positive step". Or vice versa.

File: core/builtins/src/kotlin/Progressions.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (shadrina) [reply core/builtins/src/kotlin/Progressions.kt]: Updated `generators/builtins/ranges.kt`, regenerated the `Range.kt`

File: generators/builtins/unsignedTypes.kt (hunk 3 above)
Comment (shadrina) [reply generators/builtins/unsignedTypes.kt]: Oops. Fixed 🙂

File: core/builtins/src/kotlin/Progressions.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (ilya-g) [reply core/builtins/src/kotlin/Progressions.kt]: Range.kt is the file where `ClosedRange` is defined. It's not generated, should be updated manually.

[the remark this answers is not in the document]
Comment (shadrina) [reply core/builtins/src/kotlin/Progressions.kt]: Sorry, just got it. Added docs to the `ClosedRange` with suitable wording. 

I hope I haven't missed anything else.

File: core/builtins/src/kotlin/Range.kt [hunk 4]
```diff
@@ -27,6 +27,8 @@ public interface ClosedRange<T: Comparable<T>> {
 
     /**
      * Checks whether the range is empty.
+     *
+     * The closed range is empty if its start value is greater than or equal to the end value.
```
Comment (ilya-g) [inline core/builtins/src/kotlin/Range.kt:31]: "equal to" is not correct, see the condition below.

Comment (shadrina) [reply core/builtins/src/kotlin/Range.kt]: Returned the wording as for the usual `Range`

**Review by ilya-g — APPROVED** [review]

**Pull request merged**